### PR TITLE
Eb 5748/accumulated fee to header

### PR DESCRIPTION
**Pull request by sasurobert**

### CODE REVIEW

File: process/block/postprocess/feeHandler.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+// CreateBlockStarted does the cleanup before creating a new block
+func (f *feeHandler) CreateBlockStarted() {
+	f.mut.Lock()
+	f.accumulatedFees = big.NewInt(0)
+	f.mut.Unlock()
+}
+
+// GetAccumulatedFees returns the total accumulated fees
+func (f *feeHandler) GetAccumulatedFees() *big.Int {
+	f.mut.Lock()
```
Comment (SebastianMarian) [inline process/block/postprocess/feeHandler.go]: RLock and RUnlock

File: process/interface.go [hunk 2]
```diff
@@ -168,28 +168,12 @@ type TransactionVerifier interface {
...
 // TransactionFeeHandler processes the transaction fee
 type TransactionFeeHandler interface {
+	CreateBlockStarted()
+	GetAccumulatedFees() *big.Int
 	ProcessTransactionFee(cost *big.Int)
 	IsInterfaceNil() bool
 }
 
-// SpecialAddressHandler responds with needed special addresses
-type SpecialAddressHandler interface {
```
Comment (iulianpascalau) [inline process/interface.go]: I think we will need this (or a subset) for the meta processing. Anyway, can be deleted now as we will recreate it in the final form when implementing meta processing.

File: process/mock/feeAccumulatorStub.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
+package mock
+
+import "math/big"
+
+type FeeAccumulatorStub struct {
```
Comment (iulianpascalau) [inline process/mock/feeAccumulatorStub.go]: Missing stub comment

File: process/mock/headerHandlerStub.go [hunk 4]
```diff
@@ -2,6 +2,7 @@ package mock
 
 import (
 	"github.com/ElrondNetwork/elrond-go/data"
```
Comment (iulianpascalau) [inline process/mock/headerHandlerStub.go]: imports not sorted

File: process/block/postprocess/feeHandler.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
+package postprocess
```
Comment (iulianpascalau) [inline process/block/postprocess/feeHandler.go]: struct not tested. Maybe add a task to improve code coverage on this struct, intermediateResults.go and preprocess subpackage (we have there some files with poor code coverage)

File: process/block/preprocess/rewardTxPreProcessor.go [hunk 6]
```diff
@@ -218,7 +197,7 @@ func (rtp *rewardTxPreprocessor) ProcessBlockTransactions(
 			txData := rtp.rewardTxsForBlock.txHashAndInfo[string(txHash)]
 			rtp.rewardTxsForBlock.mutTxsForBlock.RUnlock()
 			if txData == nil || txData.tx == nil {
```
Comment (iulianpascalau) [inline process/block/preprocess/rewardTxPreProcessor.go]: check.IfNil(txData.tx)

File: data/block/block.go [hunk 7]
```diff
@@ -200,6 +202,16 @@ func (h *Header) GetReceiptsHash() []byte {
 	return h.ReceiptsHash
 }
 
+// GetAccumulatedFees returns the accumulated fees in the header
+func (h *Header) GetAccumulatedFees() *big.Int {
+	return big.NewInt(0).Set(h.AccumulatedFees)
+}
+
+// SetAccumulatedFees returns the accumulated fees in the header
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply data/block/block.go]: done.

File: data/block/metaBlock.go [hunk 8]
```diff
@@ -197,6 +198,16 @@ func (m *MetaBlock) GetReceiptsHash() []byte {
 	return m.ReceiptsHash
 }
 
+// GetAccumulatedFees returns the accumulated fees in the header
+func (m *MetaBlock) GetAccumulatedFees() *big.Int {
+	return big.NewInt(0).Set(m.AccumulatedFees)
+}
+
+// SetAccumulatedFees returns the accumulated fees in the header
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply data/block/metaBlock.go]: done.

File: epochStart/genesis/genesisBlockCreator.go [hunk 9]
```diff
@@ -284,23 +285,22 @@ func createProcessorsForMetaGenesisBlock(
...
 
-	nilTxFeeHandler := &metachain.TransactionFeeHandler{}
 	txProcessor, err := processTransaction.NewMetaTxProcessor(
 		args.Accounts,
 		args.AddrConv,
 		args.ShardCoordinator,
 		scProcessor,
 		txTypeHandler,
-		nilTxFeeHandler, //we need the nil fee handler in order to process the the staking transactions
+		genesisFeeHandler, //we need the nil fee handler in order to process the the staking transactions
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply epochStart/genesis/genesisBlockCreator.go]: done.

File: process/block/postprocess/feeHandler.go (hunk 1 above)
Comment (sasurobert) [reply process/block/postprocess/feeHandler.go]: done.

File: cmd/node/factory/structs.go [hunk 10]
```diff
@@ -6,6 +6,7 @@ import (
 	"crypto/rand"
 	"encoding/hex"
 	"errors"
+	"github.com/ElrondNetwork/elrond-go/process/block/postprocess"
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply cmd/node/factory/structs.go]: done.

File: consensus/mock/headerHandlerStub.go [hunk 11]
```diff
@@ -2,6 +2,7 @@ package mock
 
 import (
 	"github.com/ElrondNetwork/elrond-go/data"
+	"math/big"
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply consensus/mock/headerHandlerStub.go]: done.

File: process/interface.go (hunk 2 above)
Comment (sasurobert) [reply process/interface.go]: not needed, it is handled directly in peer processor.

File: process/mock/feeAccumulatorStub.go (hunk 3 above)
Comment (sasurobert) [reply process/mock/feeAccumulatorStub.go]: done.

File: process/mock/headerHandlerStub.go (hunk 4 above)
Comment (sasurobert) [reply process/mock/headerHandlerStub.go]: done.

File: process/block/postprocess/feeHandler.go (hunk 5 above)
Comment (sasurobert) [reply process/block/postprocess/feeHandler.go]: agreed. Test will be added for economics V2. added task.

File: process/block/preprocess/rewardTxPreProcessor.go (hunk 6 above)
Comment (sasurobert) [reply process/block/preprocess/rewardTxPreProcessor.go]: done.

File: process/block/preprocess/rewardTxPreProcessor.go [hunk 12]
```diff
@@ -331,7 +331,6 @@ func (rtp *rewardTxPreprocessor) processRewardTransaction(
 		return err
 	}
 
-	//TODO: These lines could be deleted as in this point these values are already set (this action only overwrites them)
```
Comment (SebastianMarian) [inline process/block/preprocess/rewardTxPreProcessor.go]: I would say to let this TODO as these lines really could be deleted from all these methods: func (rtp *rewardTxPreprocessor) processRewardTransaction, func (scr *smartContractResults) processSmartContractResult( and func (txs *transactions) processAndRemoveBadTransaction(

File: process/mock/feeAccumulatorStub.go [hunk 13]
```diff
@@ -2,6 +2,7 @@ package mock
 
 import "math/big"
 
+// FeeAccumulatorSub is a stub which implement TransactionFeeHandler interface
```
Comment (SebastianMarian) [inline process/mock/feeAccumulatorStub.go]: implements

File: process/interface.go (hunk 2 above)
Comment (iulianpascalau) [reply process/interface.go]: ok

File: process/block/postprocess/feeHandler.go (hunk 5 above)
Comment (iulianpascalau) [reply process/block/postprocess/feeHandler.go]: :+1:

File: process/block/preprocess/rewardTxPreProcessor.go (hunk 12 above)
Comment (sasurobert) [reply process/block/preprocess/rewardTxPreProcessor.go]: done.

File: process/mock/feeAccumulatorStub.go (hunk 13 above)
Comment (sasurobert) [reply process/mock/feeAccumulatorStub.go]: done.

**Pull request merged**